src/units/unit.cpp

```diff
@@ -1457,8 +1457,13 @@ static bool matches_ability_filter(const config & cfg, const std::string& tag_na
 	if(!bool_matches_if_present(filter, cfg, "affect_self", true))
 		return false;
 
-	if(!bool_matches_if_present(filter, cfg, "affect_allies", true))
+	if(!filter["affect_allies"].empty() && !cfg["affect_allies"].empty()){
+		if(!bool_matches_if_present(filter, cfg, "affect_allies", true)){
+			return false;
+		}
+	} else if(!filter["affect_allies"].empty() && filter["affect_allies"].str() != "empty"){
 		return false;
+	}
```

Comment:
This is very confusing. This should check if affect_allies exists in the filter at all. Then it should check if it is empty and return the comparison to what's in cfg or if it's not empty do the check as before.

Reply:
Because when affect_allies  no specified in abilites only unit adjacent of same side affected but nt unit of allied side and if affect_allies =no none unit allies, or same side are affected affect_allies has 3 vale,true,false and empty because absence of attributechas value différente what if affect_allies =no/yes

Reply:
Wiki is correct for affect_allies  not have default_value

Reply:
> Because when affect_allies no specified in abilites only unit adjacent of same side affected but nt unit of allied side and if affect_allies =no none unit allies, or same side are affected affect_allies has 3 vale,true,false and empty because absence of attributechas value différente what if affect_allies =no/yes

Ok. Does any of that change how to properly implement the check?

Reply:
like affect_allies no have default value check if abilitie has this attribute is also important what value of attribute dand filter["filter_allies"]=empty is for check if abilitie using attribute or not.

Reply:
This is getting surreal. Do you understand the difference between an empty string with no characters and the string `empty` with 5 characters?

Reply:
then suggest another string instead of empty and i change.

Reply:
"empty" changed to "no_used"

Reply:
wtf

Reply:
> wtf

excuse me?

Reply:
As per the [wiki](https://wiki.wesnoth.org/AbilitiesWML#Common_keys_and_tags_for_every_ability), there are three valid values: yes, no, and not present.

The literal string value of `"empty"` or `"no_used"` is not valid anywhere.

Reply:
Please change it back to "empty".

Reply:
I also don't understand what @soliton- and @Pentarctagon are complaining about here; using the string "empty" for this purpose seems perfectly natural to me, or perhaps "default" would work instead. Admittedly, using the empty string for that purpose isn't terrible, I suppose. It's not like this string is going to be matched directly against the value of `affect_allies` in the abilitiy; it's an arbitrary string that means the `affect_allies` attribute **must not be present** in the ability. So whether that string is `"empty"` or `"default"` or `"no_used"` or `""` doesn't make that much of a difference.

Reply:
It makes no sense to use a different value then the ability uses. That is just confusing. The ability uses `affect_allies=` or `affect_allies=""` if that is more understandable. There is zero reason to not simply do the same in the filter.

Reply:
Ok, I misunderstood that the ability code uses the empty value instead of not specifying the attribute but the wiki clearly says "not used". My bad. If someone can point out the exact code where this is handled I would appreciate it.
https://github.com/wesnoth/wesnoth/blob/000ba55eb1c5f0261132aa734aaf4b05bb0fdf21/src/units/abilities.cpp#L1037-L1038
This handles `affect_allies` but does not treat not set special as far as I can tell.

So we do need a special value in the filter since we do need to distinguish between the filter just not filtering for affect_allies and filtering for affect_allies to not be there in the ability. "empty" or "" would be confusing for that so "not set" exactly as the wiki puts it or "not used" would be options.

Reply:
perhaps what imust added "smae_side_only" in code for could used like value by default?

Reply:
> Please change it back to "empty".

done